File: 백준/Platinum/2243. 사탕상자/사탕상자.java
import java.io.*;
import java.util.*;

public class Main {
    // 세그먼트 트리 초기화
    static int[] segTree = new int[1000001*4];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());

            if (Integer.parseInt(st.nextToken()) == 2) {
                int idx = Integer.parseInt(st.nextToken());
                int amount = Integer.parseInt(st.nextToken());
                change(0, 1000000, idx, amount, 0);
            } else {
                int rank = Integer.parseInt(st.nextToken());
                System.out.println(query(0, 1000000, rank ,0));
            }
        }
    }

    static int query(int left, int right, int rank, int Node) {
        segTree[Node]--;

        if (left == right) {
            return left;
        }

        int mid = (left + right) / 2;

        if (segTree[Node*2+1] >= rank) {
            return query(left, mid, rank, Node*2+1);
        } else {
            return query(mid+1, right, rank-segTree[Node*2+1], Node*2+2);
        }
    }

    static int change(int left, int right, int idx, int amount, int Node) {

        segTree[Node] += amount;

        if (left == right) {
            return segTree[Node];
        }

        int mid = (left + right) / 2;

        if (mid >= idx) {
            return change(left, mid, idx, amount, Node*2+1);
        } else {
            return change(mid+1, right, idx, amount, Node*2+2);
        }
    }
}

File: 백준/Platinum/2243. 사탕상자/사탕상자.py
# 사탕상자
import sys
input = sys.stdin.readline

def query(left, right, rank, Node):

    segTree[Node] -= 1
    
    if left == right:
          return left
        
    mid = (left + right) // 2

    if segTree[Node*2+1] >= rank:
        return query(left, mid, rank, Node*2+1)
    else:
        return query(mid+1, right, rank-segTree[Node*2+1], Node*2+2)

def change(left, right, idx, amount, Node):

    # 바꾸고자하는 idx 범위를 벗어난 경우
    if idx < left or idx > right:
        return segTree[Node]

    if left == right:
        segTree[Node] += amount
        return segTree[Node]

    mid = (left + right) // 2

    leftChange = change(left, mid, idx, amount, Node*2+1)
    rightChange = change(mid+1, right, idx, amount, Node*2+2)
    # 세그먼트 트리 갱신
    segTree[Node] = leftChange + rightChange

    return segTree[Node]
    
n = int(input())
segTree = [0] * (1000001*4)

for i in range(n):
    cmd = list(map(int,input().split()))

    if cmd[0] == 2:
        change(0, 1000000, cmd[1], cmd[2], 0)
    else:
        rank = cmd[1]
        result = query(0, 1000000, rank, 0)
        print(result)

